src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs

```diff
@@ -192,6 +198,16 @@ private static uint HandleEventShutdownComplete(State state, ref ConnectionEvent
 
             // Stop accepting new streams.
             state.AcceptQueue.Writer.Complete();
+
+            // Stop notifying about available streams.
+            lock (state)
+            {
+                state.NewUnidirectionalStreamsAvailable?.SetException(ExceptionDispatchInfo.SetCurrentStackTrace(new QuicOperationAbortedException()));
```

Comment:
I remembered that in Quic we always had SetException and SetResult outside of the lock:
https://github.com/dotnet/runtime/blob/5fc93f425ead47546efcbe8fa1be0b1becc03c25/src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs#L394-L407
I wonder why is it like that and whether that's a best practice? @scalablecory @geoffkizer

Reply:
I pulled the `SetException` outside of the lock. What is important is that the swaps of the variables stay under the lock, the set itself doesn't have to be.